level_handler.go

```diff
@@ -224,23 +224,50 @@ func (s *levelHandler) close() error {
 	return errors.Wrap(err, "levelHandler.close")
 }
 
-// getTableForKey acquires a read-lock to access s.tables. It returns a list of tableHandlers.
-func (s *levelHandler) getTableForKey(key []byte) []*table.Table {
+// refTablesForKey acquires a read-lock to access s.tables. It returns a list of tableHandlers.
+func (s *levelHandler) refTablesForKey(key []byte) []*table.Table {
 	s.RLock()
 	defer s.RUnlock()
 
 	if s.level == 0 {
-		// For level 0, we need to check every table. Remember to make a copy as s.tables may change
-		// once we exit this function, and we don't want to lock s.tables while seeking in tables.
-		// CAUTION: Reverse the tables.
-		out := make([]*table.Table, 0, len(s.tables))
-		for i := len(s.tables) - 1; i >= 0; i-- {
-			out = append(out, s.tables[i])
-			s.tables[i].IncrRef()
-		}
+		return s.refLevel0Tables()
+	}
+	tbl := s.refLevelNTable(key)
+	if tbl == nil {
+		return nil
+	}
+	return []*table.Table{tbl}
+}
+
+// refTablesForKeys return tables for pairs.
+// level0 returns all tables.
+// level1+ returns tables for every key.
+func (s *levelHandler) refTablesForKeys(pairs []keyValuePair) []*table.Table {
+	s.RLock()
+	defer s.RUnlock()
+	if s.level == 0 {
+		return s.refLevel0Tables()
+	}
+	out := make([]*table.Table, len(pairs))
+	for i, pair := range pairs {
```

Comment:
There are some redundant tables, these tables may cause unnecessary search.
We should group by level, table, and keys. Also we do a lot of redundant work, such as recalculate key's hash when do bloom filter looking up.

Reply:
We can do more optimization in further PR.